### ISSUE TITLE
[solved] HMAC mismatch - correct password not working because key file not selected

### ISSUE THREAD

**Issue by MaximilianKohler** [keepassxreboot/keepassxc#11462]
## Overview

I'm seeing a ton of people reporting this so I am sure it's bug. I'm really disappointed to see that in most of the issues I clicked through there wasn't even an attempt to debug the issue. Fortunately, I **may** have narrowed down the issue to bitlocker encryption or general drive/file encryption.

I use keepassxc at home. It's installed on a bitlocker drive. I backed it up to a USB thumb drive and when opening the database from that USB drive I get the error. I saw your responses sometimes asked if they were using a USB thumbdrive, and then in one issue you said it's a thumbdrive corruption issue. Thankfully, I have a copy of the database in the cloud so I downloaded it and I get the same error, so it's **not a USB drive issue**.

The other primary suspect is therefore bitlocker encryption. I searched the issues for info on this and didn't find any.

I can try to help you debug this but I don't currently have access to my desktop computer.

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Keep db on a bitlocker-protected drive
2. Copy the db to another computer or thumbdrive
3. HMAC error, invalid credentials

## Context
I use a password manager to login, so it's not possible that my password is wrong.

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - Version 2.7.9
Revision: 8f6dd13

Qt 5.15.11
Debugging mode is disabled.

Operating system: Windows 10 Version 2009
CPU architecture: x86_64
Kernel: winnt 10.0.19045

Enabled extensions:
- Auto-Type
- Browser Integration
- Passkeys
- SSH Agent
- KeeShare
- YubiKey
- Quick Unlock

Cryptographic libraries:
- Botan 3.1.1

**Comment by droidmonkey** [comment]
> I use a password manager to login, so it's not possible that my password is wrong.

Please do explain. This is the cause of all of these issues, wrong credentials.

> copy of the database in the cloud so I downloaded it and I get the same error

Definitely wrong credentials. 

I have used bitlocker for over 10 years without issue, and so do millions of people. 100% certain that has nothing to do with your issue. 

There is no way to "debug" these issues because it is impossible to tell if you have the wrong password / key file or if there is actual file corruption. If you have actual file corruption, it is much more likely to have a different error such as header size mismatch, invalid database file, or similar.

**Comment by MaximilianKohler** [comment]
> Please do explain. This is the cause of all of these issues, wrong credentials.

I'm using keepassxc only for TOTP. The password for my keepassxc databases is stored in another password manager. It has not changed. When I get back to my desktop I'm going to be able to login to the same databases with the same password. For some reason, moving the databases to another computer/drive has messed something up, and it's not because the USB flash drive is corrupted or has corrupted the database files.

**Comment by droidmonkey** [comment]
The only other explanation is faulty RAM

**Comment by MaximilianKohler** [comment]
Yet nothing else has gone wrong -- there's no sign of any problem with my RAM. Yet there are a dozen or more people reporting this HMAC mismatch issue with keepassxc. I've had zero such issues with Keepass (non-xc) for many years.

Are there no troubleshooting steps you can think of that might help me narrow down the problem? I use FreeFileSync to sync/move the files, if that matters.

**Comment by droidmonkey** [comment]
Dozens of people out of (est) 2 Million users. Of those dozens, 80% finally admitted to having the wrong password, 10% bad RAM, 10% ghosted. 

You can easily troubleshoot on your end. Use the same database on a different device. If it unlocks then it's your device. If it doesn't, then most likely it's bad password. 

> there's no sign of any problem with my RAM

There is no real way to tell without running a formal memtest. The way Argon2 works, more RAM is used than normal and requires absolute precision. If you set your decryption time very high then that would increase the chances of problems with faulty RAM.

**Comment by MaximilianKohler** [comment]
On my laptop:

* I ran Windows Memory Diagnostic, and memtest. No errors.
* I downloaded the whole Keepassxc folder from a cloud backup, thus bypassing the USB thumb drive completely for both the application and db. Same error.

On a 3rd PC:

* I tried the Keepassxc on the USB, and I tried downloading the db from a cloud backup. Same error.

You mentioned a "decryption time" setting. I looked through the settings and didn't see anything like that, so I guess it's something you select when creating a db and thus there's nothing I can change now.

So from what I can tell, you are concluding this is a bad password issue, so I'll have to wait till I get home and confirm the same password, which hasn't been changed, still works on the db on my home PC. Hopefully there will be further troubleshooting if that's the case.

**Comment by MaximilianKohler** [comment]
Well, it turns out it's more of a UI issue I would say.

1. I forgot the db had a key file, and/or I forgot to change the path (since new PC/USB).

2. I'm used to Keepass's login screen where both reminds you that there's a key, AND you can easily notice that the path needs to be changed:
![Image](https://github.com/user-attachments/assets/3a97efa5-80d4-4f3d-99c3-450b8366a9b2)

Fine, XC's method is slightly more secure/secretive, but it can probably be improved to avoid scenarios like this one.

3. When I was logging in the most recent time it looked like this (no key file selected):
![Image](https://github.com/user-attachments/assets/16e0f1c4-e99a-49de-8c29-71019e5cd496)

Now it looks like this (key file selected, but the area is compressed so you can't even tell what it is, or what the path is):
![Image](https://github.com/user-attachments/assets/b49f5c4e-7cc4-4d6d-b590-cabf8cf32c0a)

So, at minimum that visual compression bug should be fixed. But I'd really like there to be some reminder/hint about the key file and/or path (Eg: it's no longer at C:/etc).

**Comment by droidmonkey** [comment]
Hilarious. 

https://github.com/keepassxreboot/keepassxc/issues/10863

Resize your window to be taller to fix the visual glitch. Known bug right now. 

PS - we have no clue you have a key file. KeePass doesn't either... until you actually use one... same with us. If you forget you use a key file that's on you, not the UI.

**Comment by MaximilianKohler** [comment]
> we have no clue you have a key file. KeePass doesn't either... until you actually use one... same with us. If you forget you use a key file that's on you, not the UI.

This is not correct. You can see in the screenshots that it is "remembered" somewhere, for both keepass and XC. Thus, if I'm not moving PCs I don't have to do anything other than input my password. Both programs remember that I'm using a keyfile, and the location of it, and automatically keep it associated with the DB.

Therefore, my routine for a year was only entering the password to open it. Then I leave home and all of a sudden the same routine doesn't work, and there's no clue about the key file or path. Surely you'd agree that something can be improved to remedy this situation other than fixing the visual bug?

As noted, Keepass gives two clues as to what could be wrong, whereas XC just says "_invalid credentials; your DB may be corrupt_", which is much more misleading than helpful.

**Comment by phoerious** [comment]
There is absolutely nothing in the database that tells us whether it needs a key file or not. The program just remembers whether you used one previously. This information is stored in the settings on your computer and isn't transferred with your database.

**Comment by MaximilianKohler** [comment]
>This information is stored in the settings on your computer and isn't transferred with your database.

That's fine. I transferred the whole folder, settings and all. I'm using the portableapps.com version.

One easy improvement may be to not obfuscate the key file path by default. The path to the DB isn't obfuscated, but you can see in the screenshot that it acts like the key file path is a password. Is that level of secrecy really necessary? If you think it is, what if it obfuscated most of the path but not the drive (IE: show C:/)?

I would also recommend changing the error text from _"invalid credentials; your DB may be corrupt"_ to include something like "if you're using a key file, check the path is correct".

Because the current error text made me think _"well my password is correct, so let me go check github to see why my db got corrupted"_.

**Comment by phoerious** [comment]
> That's fine. I transferred the whole folder, settings and all. I'm using the portableapps.com version.

There's a second config for device-specific settings that isn't stored in your apps folder. The key file path could be an entirely different one on each of your devices.

> I would also recommend changing the error text from "invalid credentials; your DB may be corrupt" to include something like "if you're using a key file, check the path is correct".

Sorry, but we're not gonna do that. Using a key file is somewhat of a niche thing for most users and we're not going to overcomplicate the error message with all the different things that could potentially go wrong.

**Comment by MaximilianKohler** [comment]
>The key file path could be an entirely different one on each of your devices.

Yes, that's why I highlighted that when I open Keepass (non-XC) on a new PC it shows me the last path I was using, thus I can see that it needs to be changed.

I wasn't aware that key files were niche, but I understand your rejection of my wording. It was just a general idea to get started on. I'm sure you can think of a better way to clarify that a corrupted DB is not the most likely problem.

As noted, there are various ways this UX could be improved.

**Comment by MaximilianKohler** [comment]
I just tested moving the PortableApps.com folder to another location on the same drive. KeepassPro remembers the relative location of the key file **and** database. So if you store the key file somewhere in the same directory you're moving then you don't have to change anything when you move the PortableApps folder (to a new PC, thumb drive, etc.).

Whereas, KeepassXC doesn't even remember you were using a keyfile, much less the location of it. KeepassXC only remembers that you were using a keyfile if don't move anything. It does dynamically remember & change the DB path though, so you don't need to select a new DB location when you move it.

I don't know how much of this has to do with KeepassXC vs PortableApps. I'm checking with them.

**Comment by droidmonkey** [comment]
Do not use portableapps. Absolutely not supported by us.

**Comment by MaximilianKohler** [comment]
Well I just tried the non-portableapps version and it handles this even worse. It won't even open the DB if you have the DB in the same folder that you move to another location.

**Comment by IssuesPlease** [comment]
@MaximilianKohler  Just wanted to add to the discussion you started. I've been using KeePassXC for years on Windows, Mac, & Linux without a problem. Two days ago, I decided, finally, to add a keyfile (created in KPXC) to improve security. I stored one copy on an external thumbdrive, another copy in an encrypted folder in a privacy cloud, and the final copy on my Mac OS. I never moved the keyfiles after distributing them. My master PW is saved in another offline password manager in case I ever forget it. The same day I created/started using the KPXC-created keyfile stored on my Mac (very easy to identify, thanks to the KPXC suffix), I started getting corruption error messages. I was sure I was entering the right master PW, but kept getting the error message, locking me out of my PW database.

I tried using the keyfile on my thumbdrive & that one worked. Once in my PW database, I immediately disabled the keyfile. I don't know what happened, but simply **_that_** it happened was enough to frighten me into action.  I bought two security keys to use with my master password and hope they'll be more reliable. I'm also investing in mirroring my PW database (minus the most sensitive PWs) in another PW management protocol also secured with security keys. 

Just wanted to corroborate what I've read about potential problems with KPXC and keyfiles. Getting locked out of your passwords can be catastrophic. And it's little comfort to read that most people have never had this problem. Best, I think, to have different database copies on different platforms just in case. Best of luck to everybody--and thanks, @MaximilianKohler, for creating this thread to warn others of what might happen.
